vms/avm/vm.go

```diff
@@ -483,26 +483,85 @@ func (vm *VM) GetUTXOs(
 		limit = maxUTXOsToFetch
 	}
 
-	seen := ids.Set{} // IDs of UTXOs already in the list
+	if paginate {
+		return vm.getPaginatedUTXOs(addrs, startAddr, startUTXOID, limit)
+	}
+	return vm.getAllUTXOs(addrs, startAddr, startUTXOID)
+}
+
+func (vm *VM) getPaginatedUTXOs(addrs ids.ShortSet,
+	startAddr ids.ShortID,
+	startUTXOID ids.ID,
+	limit int,
+) ([]*avax.UTXO, ids.ShortID, ids.ID, error) {
+	seen := ids.New(limit) // IDs of UTXOs already in the list
 	utxos := make([]*avax.UTXO, 0, limit)
 	lastAddr := ids.ShortEmpty
 	lastIndex := ids.Empty
 	addrsList := addrs.List()
 	ids.SortShortIDs(addrsList)
+
+	for _, addr := range addrsList {
+		start := ids.Empty
+		if comp := bytes.Compare(addr.Bytes(), startAddr.Bytes()); comp == -1 { // Skip addresses before [startAddr]
+			continue
+		} else if comp == 0 {
+			start = startUTXOID
+		}
+
+		utxoIDs, err := vm.state.Funds(addr.Bytes(), start, limit) // Get UTXOs associated with [addr]
+		if err != nil {
+			return nil, ids.ShortID{}, ids.ID{}, fmt.Errorf("couldn't get UTXOs for address %s: %w", addr, err)
+		}
+		for _, utxoID := range utxoIDs {
+			if seen.Contains(utxoID) { // Already have this UTXO in the list
+				continue
+			}
+			utxo, err := vm.state.UTXO(utxoID)
+			if err != nil {
+				return nil, ids.ShortID{}, ids.ID{}, fmt.Errorf("couldn't get UTXO %s: %w", utxoID, err)
+			}
+			utxos = append(utxos, utxo)
+			seen.Add(utxoID)
+			lastAddr = addr
+			lastIndex = utxoID
+			limit--
+			if limit <= 0 {
+				break // Found [limit] utxos; stop.
+			}
+		}
+	}
+	return utxos, lastAddr, lastIndex, nil
+}
+
+func (vm *VM) getAllUTXOs(addrs ids.ShortSet,
+	startAddr ids.ShortID,
+	startUTXOID ids.ID,
+) ([]*avax.UTXO, ids.ShortID, ids.ID, error) {
```

Comment:
Similar comment to Aaron's, but I'm not sure what information the `lastAddr` and `lastIndex` really convey in this implementation now (As they were previously used for pagination, but that is no longer relevant here)

Reply:
I might be misunderstanding the underlying condition, but we'll be losing the ability to get all UTXOs at a set address/utoxid. Happy to remove that if it's not an issue :)

Reply:
I was talking about the returned values here, not necessarily the arguments.